crates/wasi/src/preview2/host/clocks.rs

```diff
@@ -52,30 +52,57 @@ impl<T: WasiView> monotonic_clock::Host for T {
         Ok(self.ctx().monotonic_clock.resolution())
     }
 
-    fn subscribe(&mut self, when: Instant, absolute: bool) -> anyhow::Result<Resource<Pollable>> {
+    fn subscribe_instant(&mut self, when: Instant) -> anyhow::Result<Resource<Pollable>> {
         let clock_now = self.ctx().monotonic_clock.now();
-        let duration = if absolute {
-            Duration::from_nanos(when - clock_now)
+        let deadline = if when <= clock_now {
+            let duration = Duration::from_nanos(when - clock_now);
+            tokio::time::Instant::now()
+                .checked_add(duration)
+                // This should only reachable if we are running on a platform
+                // where std::time::Instant (of which tokio's Instant is a
+                // newtype wrapper) represents less than a u64 in nanoseconds,
+                // which appears to always be true in practice, but isnt a
+                // guarantee provided by std.
+                .ok_or_else(|| anyhow::anyhow!("time overflow: duration {duration:?}"))?
         } else {
-            Duration::from_nanos(when)
+            tokio::time::Instant::now()
         };
-        let deadline = tokio::time::Instant::now()
-            .checked_add(duration)
-            .ok_or_else(|| anyhow::anyhow!("time overflow: duration {duration:?}"))?;
         // NB: this resource created here is not actually exposed to wasm, it's
         // only an internal implementation detail used to match the signature
         // expected by `subscribe`.
-        let sleep = self.table_mut().push(Sleep(deadline))?;
+        let sleep = self.table_mut().push(Deadline::Instant(deadline))?;
+        subscribe(self.table_mut(), sleep)
+    }
```

Comment:
Could this method perhaps delegate to `subscribe_duration` or a similar helper once it calculates the relative desired `Duration`? Or alternatively would it be possible to directly convert between `when` and a `tokio::time::Instant`?

It basically felt a bit odf that the `Never` case was handled below but not here

Reply:
Thanks, I thought that over again and factored it out into a helper function.